core/Symbols.cc

```diff
@@ -1112,6 +1112,116 @@ TypePtr Symbol::sealedSubclassesToUnion(const GlobalState &gs) const {
+            ENFORCE(isa_type<ClassType>(elem), "Something in requiredAncestors that's not a ClassType");
+            if (cast_type_nonnull<ClassType>(elem).symbol == ancestor) {
+                return;
+            }
+        }
+    }
+
+    auto type = core::make_type<ClassType>(ancestor);
+    (cast_type<TupleType>(ancestors.data(ctx)->resultType))->elems.emplace_back(type);
+    ancestors.data(ctx)->locs_.emplace_back(loc);
+}
+
+// Locally required ancestors by this class or module
+vector<Symbol::RequiredAncestor> Symbol::requiredAncestors(const GlobalState &gs) const {
+    ENFORCE(this->isClassOrModule(), "Symbol is not a class or module: {}", this->show(gs));
+
+    vector<RequiredAncestor> res;
+
+    auto ancestors = this->findMember(gs, Names::requiredAncestors());
+    if (!ancestors.exists()) {
+        // No ancestor was recorded for this class or module
+        return res;
+    }
+
+    auto data = ancestors.data(gs);
+    auto types = cast_type<TupleType>(data->resultType);
+    auto index = 0;
+    for (auto elem : types->elems) {
+        ENFORCE(isa_type<ClassType>(elem), "Something in requiredAncestors that's not a ClassType");
+        ENFORCE(index < data->locs().size(), "Missing loc in requiredAncestors");
+
+        RequiredAncestor req;
+        req.symbol = cast_type_nonnull<ClassType>(elem).symbol;
+        req.loc = data->locs().at(index);
+        res.emplace_back(req);
+
+        index++;
+    }
+
+    return res;
+}
+
+// All required ancestors by this class or module
+vector<Symbol::RequiredAncestor> requiredAncestorsTransitiveInternal(const GlobalState &gs, SymbolRef sym,
+                                                                     std::vector<SymbolRef> seen) {
+    std::vector<Symbol::RequiredAncestor> res;
+
+    if (std::find(seen.begin(), seen.end(), sym) != seen.end()) {
+        return res; // Break recursive loops if we already visited this ancestor
+    }
+    seen.emplace_back(sym);
+
+    for (auto ancst : sym.data(gs)->requiredAncestors(gs)) {
+        ancst.origin = sym;
+        res.emplace_back(ancst);
+        for (auto sancst : requiredAncestorsTransitiveInternal(gs, ancst.symbol, seen)) {
+            if (sancst.symbol == sym) {
+                continue;
+            }
+            res.emplace_back(sancst);
+        }
+    }
+
+    auto parent = sym.data(gs)->superClass();
+    if (parent.exists()) {
+        for (auto ancst : requiredAncestorsTransitiveInternal(gs, parent, seen)) {
+            if (ancst.symbol == sym) {
+                continue;
+            }
+            res.emplace_back(ancst);
+        }
+    }
+
+    for (auto mixin : sym.data(gs)->mixins()) {
+        for (auto ancst : mixin.data(gs)->requiredAncestors(gs)) {
+            if (ancst.symbol == sym) {
+                continue;
+            }
+            ancst.origin = mixin;
+            res.emplace_back(ancst);
+        }
+    }
+
+    return res;
```

Comment:
Ideally we should cache this result the first time it's called, but we not might have a mutable context at the time we will call this method (ie in `calls.cc`).